src/Microsoft.ApiDesignGuidelines.Analyzers/Core/DoNotHideBaseClassMethods.cs

```diff
@@ -0,0 +1,100 @@
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Design,
+            DiagnosticSeverity.Warning, 
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182143.aspx",
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.RegisterSymbolAction(SymbolAnalyzer, SymbolKind.Method);
+        }
+
+        private void SymbolAnalyzer(SymbolAnalysisContext context)
+        {
+            var method = (IMethodSymbol)context.Symbol;
+            if (method.IsOverride)  // Since this is common, check for it now
+            {
+                return;
+            }
+
+            foreach (var hiddenMethod in GetMethodsHiddenByMethod(method, method.ContainingType.BaseType))
+            {
+                var diagnostic = Diagnostic.Create(Rule, context.Symbol.Locations[0], method.ToDisplayString(), hiddenMethod.ToDisplayString());
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private IEnumerable<IMethodSymbol> GetMethodsHiddenByMethod(IMethodSymbol method, INamedTypeSymbol baseType)
+        {
+            while (true)
+            {
+                if (baseType?.BaseType == null)
+                {
+                    // There are no other base types to check - we're at the top of the hierarchy
+                    yield break;
+                }
+
+                var baseTypeMethods = baseType.GetMembers(method.Name)
+                    .Where(x => !x.IsStatic && !x.IsVirtual && x.DeclaredAccessibility != Accessibility.Private)
+                    .OfType<IMethodSymbol>();
+
+                foreach (var baseTypeMethod in baseTypeMethods)
+                {
+                    if (method.Parameters.Length != baseTypeMethod.Parameters.Length)
+                    {
+                        continue;
+                    }
+
+                    var zippedParams = method.Parameters.Zip(
+                        baseTypeMethod.Parameters, 
+                        (x, y) => new {DerivedParm = x, BaseParam = y}).ToList();
+
+                    // Does the matching base type method also have the same return type, the same parameter names, 
+                    // and at least one parameter with a type that is more derived than the derived type's corresponding parameter? 
+                    if (method.ReturnType.Equals(baseTypeMethod.ReturnType) 
+                        && zippedParams.All(x => x.DerivedParm.Name == x.BaseParam.Name) 
+                        && zippedParams.Any(x => !x.BaseParam.Type.Equals(x.DerivedParm.Type) 
+                        && x.BaseParam.Type.DerivesFrom(x.DerivedParm.Type)))
```

Comment:
What if we have following: Base.Method(Derived d, Type1 t) and Derived.Method(Base b, Type2 t), where Type1 and Type2 are unrelated with no conversion between each other. We should not fire, but now we will fire? Can you add tests?